### PR TITLE
[sml] add Gaussian naive bayes in jax

### PR DESCRIPTION
**Pull request by nantsing**
使用 SPU 实现 Gaussian naive bayes 算法

Issue Number:  #256 

1. 完成了fit, partial_fit, predict 基础功能
2. 提供了var_smoothing 参数
3. emulation 和 test 实验中，在 iris, wine, 随机数据集上与 sklearn.naive_bayes.GaussianNB 表现均一致

说明：
与 sklearn 不一样的地方在于，在创建分类器时，需要首先提供数据可能出现的所有类别 (classes_ 参数)

### CODE REVIEW

**Comment by nantsing** [comment]
***I have read the CLA Document and I hereby sign the CLA***

**Comment by nantsing** [comment]
**recheck**

**Comment by nantsing** [comment]
> **CLA Assistant Lite bot:** Thank you for your submission, we really appreciate it. Like many open-source projects, we ask that you sign our [Contributor License Agreement](https://github.com/secretflow/.github/blob/main/CLA.md) before we can accept your contribution. You can sign the CLA by just posting a Pull Request Comment same as the below format.
> 
> _**I have read the CLA Document and I hereby sign the CLA**_
> 
> You can retrigger this bot by commenting **recheck** in this Pull Request

**rechck**

**Comment by nantsing** [comment]
> **CLA Assistant Lite bot:** Thank you for your submission, we really appreciate it. Like many open-source projects, we ask that you sign our [Contributor License Agreement](https://github.com/secretflow/.github/blob/main/CLA.md) before we can accept your contribution. You can sign the CLA by just posting a Pull Request Comment same as the below format.
> 
> _**I have read the CLA Document and I hereby sign the CLA**_
> 
> You can retrigger this bot by commenting **recheck** in this Pull Request

**recheck**

**Comment by nantsing** [comment]
> **CLA Assistant Lite bot:** Thank you for your submission, we really appreciate it. Like many open-source projects, we ask that you sign our [Contributor License Agreement](https://github.com/secretflow/.github/blob/main/CLA.md) before we can accept your contribution. You can sign the CLA by just posting a Pull Request Comment same as the below format.
> 
> _**I have read the CLA Document and I hereby sign the CLA**_
> 
> You can retrigger this bot by commenting **recheck** in this Pull Request

recheck

**Comment by nantsing** [comment]
recheck

**Comment by deadlywing** [comment]
@nantsing 

Hi, 麻烦确保commit的python代码已经被black格式化，bazel files被buildfier格式化哈～

File: sml/naive_bayes/emulations/gnb_emul.py [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+            n_samples=n_samples, n_features=n_features, centers=centers
+        )
+        classes = jnp.unique(y)
+        assert len(classes) == centers, f'Retry or increase partial.'
+        total_samples = len(y)
+        split_idx = int(partial * len(y))
+        X1, y1 = X[:split_idx], y[:split_idx]
+        X2, y2 = X[split_idx:], y[split_idx:]
+
+        y1_pred, y2_pred = emulator.run(proc_fit)(X1, y1, X2, y2, classes)
```
Comment (deadlywing) [inline sml/naive_bayes/emulations/gnb_emul.py:66]: emulator.run的输入数据需要是被 `emulator.seal` 处理过的数据哈～
具体可以参考其余emulations内的代码～

**Comment by deadlywing** [comment]
PS：想问一下是否必须使用FM128？FM64会造成巨大的精度drop么？

**Comment by nantsing** [comment]
> PS：想问一下是否必须使用FM128？FM64会造成巨大的精度drop么？

sf 默认时FM64 吗，我没有改过这个参数，从测试结果来看好像不会有很大的精度问题。在我自己的测试中算法表现和sklearn差不多的。

File: sml/naive_bayes/emulations/gnb_emul.py (hunk 1 above)
Comment (nantsing) [reply sml/naive_bayes/emulations/gnb_emul.py]: 这里改好了，black 和 buildifier也确认做过一次了

**Comment by nantsing** [comment]
image

想问一下图中的报错信息是指我要将第30行代码改为第31行所示吗？但我 black 完之后仍然是30行的写法

**Comment by nantsing** [comment]
> image
> 想问一下图中的报错信息是指我要将第30行代码改为第31行所示吗？但我 black 完之后仍然是30行的写法

@deadlywing

**Comment by deadlywing** [comment]
Hi，我本地试了一下，black后是这样的。您是用的vscode插件么？

image

**Comment by nantsing** [comment]
> Hi，我本地试了一下，black后是这样的。您是用的vscode插件么？
> 
> image

我是在 wsl 的终端运行的 black 命令。刚刚换到 MacOS 终端运行之后确实是第31行这种写法，感觉有点奇怪了

**Comment by deadlywing** [comment]
> > Hi，我本地试了一下，black后是这样的。您是用的vscode插件么？
> > image
> 
> 我是在 wsl 的终端运行的 black 命令。刚刚换到 MacOS 终端运行之后确实是第31行这种写法，感觉有点奇怪了

@anakinxc 
有啥想法不，是windons和类unix系统black的style有区别么？

**Comment by tpppppub** [comment]
> > > Hi，我本地试了一下，black后是这样的。您是用的vscode插件么？
> > > image
> > 
> > 
> > 我是在 wsl 的终端运行的 black 命令。刚刚换到 MacOS 终端运行之后确实是第31行这种写法，感觉有点奇怪了
> 
> @anakinxc 有啥想法不，是windons和类unix系统black的style有区别么？

black 版本不一样吧

File: sml/naive_bayes/gnb.py [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+
+        Returns
+        -------
+        jll: array-like of shape (n_samples, n_classes)
+            The joint log likelihood of testing samples w.r.t all classes.
+        """
+        joint_log_likelihood = []
+        for i in range(len(self.classes_)):
+            jointi = jnp.where(
+                self.class_prior_[i] != 0, jnp.log(self.class_prior_[i]), -jnp.inf
```
Comment (deadlywing) [inline sml/naive_bayes/gnb.py:244]: 建议comment一下，这里的`jnp.where`是必须的，原因是spu计算的 `log(0) ` 和 `-jnp.inf` 这种字面量是不一样的

File: sml/naive_bayes/emulations/3pc.json
Comment (deadlywing) [inline sml/naive_bayes/emulations/3pc.json:1]: 这个配置文件不必要

File: sml/naive_bayes/tests/gnb_test.py [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import os
+import sys
+import unittest
+
+import jax.numpy as jnp
+import numpy as np
+from jax import random
```
Comment (deadlywing) [inline sml/naive_bayes/tests/gnb_test.py:21]: clean up 没用的import

File: sml/naive_bayes/emulations/gnb_emul.py [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import os
+import sys
+
+import jax.numpy as jnp
+import jax.random as random
```
Comment (deadlywing) [inline sml/naive_bayes/emulations/gnb_emul.py:19]: clean up没用的import

File: sml/naive_bayes/tests/gnb_test.py [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+        def proc(X1, y1, X2, y2, classes):
+            model = GaussianNB(
+                classes_=classes,
+                var_smoothing=1e-6,
+            )
+
+            model.fit(X1, y1)
+            y1_pred = model.predict(X)
+
+            model.fit(X2, y2)
```
Comment (deadlywing) [inline sml/naive_bayes/tests/gnb_test.py:63]: 应该调用`partial_fit`吧

File: sml/naive_bayes/tests/gnb_test.py [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+            y2_pred = model.predict(X)
+
+            return y1_pred, y2_pred
+
+        def test_precision(X1, y1, X2, y2, classes):
```
Comment (deadlywing) [inline sml/naive_bayes/tests/gnb_test.py:68]: 没必要再跑一次，theta和var可以在上面那个函数里一起返回

File: sml/naive_bayes/gnb.py [hunk 7]
```diff
@@ -0,0 +1,266 @@
...
+        # Combine mean of old and new data.
+        total_mu = jnp.where(n_new == 0.0, mu, (n_new * new_mu + n_past * mu) / n_total)
+
+        # Combine variance of old and new data.
+        old_ssd = n_past * var
+        new_ssd = n_new * new_var
+        total_ssd = old_ssd + new_ssd + (n_new * n_past / n_total) * (mu - new_mu) ** 2
+        total_var = jnp.where(n_new == 0.0, var, total_ssd / n_total)
+        # correct precision problem
+        total_var = jnp.where(total_var < 0, 0.0, total_var)
```
[the remark this answers is not in the document]
Comment (nantsing) [reply sml/naive_bayes/gnb.py:167]: 这里是当 var 本该为 0 时，由于精度问题 var 可能会为负

File: sml/naive_bayes/gnb.py [hunk 8]
```diff
@@ -0,0 +1,266 @@
...
+            )
+
+            self.class_count_ = self.class_count_.at[i].set(self.class_count_[i] + N_i)
+
+            for j in range(n_features):
+                self.theta_ = self.theta_.at[i, j].set(new_theta[j])
+                self.var_ = self.var_.at[i, j].set(new_sigma[j])
+
+        self.var_ = self.var_ + self.epsilon_
+        self.class_prior_ = self.class_count_ / jnp.sum(self.class_count_)
```
[the remark this answers is not in the document]
Comment (nantsing) [reply sml/naive_bayes/gnb.py:206]: 是当前所有的样本数，因为有 partial_fit，后面可能还会有新的样本。用明文替代这个密文具体是什么意思呢？

File: sml/naive_bayes/gnb.py [hunk 9]
```diff
@@ -0,0 +1,266 @@
...
+        n_new = N_i
+        new_mu = jnp.sum(X, axis=0) / n_new
+        new_var = jnp.sum((X - new_mu) ** 2, axis=0)
+        new_var = new_var - N * new_mu**2 + N_i * new_mu**2
+        new_var = new_var / n_new
+
+        n_total = n_past + n_new
+
+        # Combine mean of old and new data.
+        total_mu = jnp.where(n_new == 0.0, mu, (n_new * new_mu + n_past * mu) / n_total)
```
[the remark this answers is not in the document]
Comment (nantsing) [reply sml/naive_bayes/gnb.py:159]: 不过按照 sklearn 的要求，也不会出现 n_total == 0 的情况

File: sml/naive_bayes/gnb.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (deadlywing) [reply sml/naive_bayes/gnb.py:167]: 是测试的时候遇到过var<0的情况么？

File: sml/naive_bayes/gnb.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (nantsing) [reply sml/naive_bayes/gnb.py:159]: 因为我的实现可以处理某一类没有数据的情况，所以加了这个jnp.where。（总类别是需要用户提供的）

[the remark this answers is not in the document]
Comment (deadlywing) [reply sml/naive_bayes/gnb.py:159]: 如果要考虑除0风险的话，`new_mu = jnp.sum(X, axis=0) / n_new`这里就已经有问题了

[the remark this answers is not in the document]
Comment (deadlywing) [reply sml/naive_bayes/gnb.py:159]: 一般如果某一个class的出现次数为0，就不会把这个class作为备选项了吧.

File: sml/naive_bayes/gnb.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (deadlywing) [reply sml/naive_bayes/gnb.py:206]: 即使有后续的`partial_fit`，总样本数也是预先可知的（每次`partial_fit`的时候累积）；

用明文替代密文：就是用上面我说的可预知的当前总样本数(明文)来替代`jnp.sum(self.class_count_)`（密文），这样可以减少一次密态除

File: sml/naive_bayes/gnb.py [hunk 10]
```diff
@@ -0,0 +1,266 @@
...
+            Target values.
+
+        Returns
+        -------
+        self : object
+        """
+        classes = self.classes_
+        n_features = self.n_features
+        y_ = jnp.expand_dims(y, 1).repeat(n_features, 1)
+        for i, y_i in enumerate(classes):
```
Comment (deadlywing) [inline sml/naive_bayes/gnb.py:190]: 这个循环可以用vamp替代

File: sml/naive_bayes/gnb.py [hunk 11]
```diff
@@ -0,0 +1,266 @@
...
+        X : array-like of shape (n_samples, n_features)
+            Testing vectors.
+
+        Returns
+        -------
+        jll: array-like of shape (n_samples, n_classes)
+            The joint log likelihood of testing samples w.r.t all classes.
+        """
+        joint_log_likelihood = []
+        for i in range(len(self.classes_)):
```
Comment (deadlywing) [inline sml/naive_bayes/gnb.py:242]: 同理，这里的循环也能用vmap替代

**Comment by deadlywing** [comment]
vmap替代，我举一个例子，另一个可以对应的修改一下
```python
def _joint_log_likelihood_single(self, prior, theta, var, X):
        jointi = jnp.where(prior != 0, jnp.log(prior), -jnp.inf)
        n_ij = -0.5 * jnp.sum(jnp.log(2.0 * jnp.pi * var))
        n_ij -= 0.5 * jnp.sum(((X - theta) ** 2) / var, 1)
        return jointi + n_ij

    def _joint_log_likelihood(self, X):
        """Compute the log likehood.

        Parameters
        ----------
        X : array-like of shape (n_samples, n_features)
            Testing vectors.

        Returns
        -------
        jll: array-like of shape (n_samples, n_classes)
            The joint log likelihood of testing samples w.r.t all classes.
        """
        # joint_log_likelihood = []
        # for i in range(len(self.classes_)):
        #     jointi = jnp.where(
        #         self.class_prior_[i] != 0, jnp.log(self.class_prior_[i]), -jnp.inf
        #     )
        #     n_ij = -0.5 * jnp.sum(jnp.log(2.0 * jnp.pi * self.var_[i, :]))
        #     n_ij -= 0.5 * jnp.sum(((X - self.theta_[i, :]) ** 2) / self.var_[i, :], 1)
        #     joint_log_likelihood.append(jointi + n_ij)
        joint_log_likelihood = vmap(
            self._joint_log_likelihood_single, in_axes=(0, 0, 0, None)
        )(self.class_prior_, self.theta_, self.var_, X)

        joint_log_likelihood = jnp.array(joint_log_likelihood).T
        return joint_log_likelihood
```

File: sml/naive_bayes/gnb.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (nantsing) [reply sml/naive_bayes/gnb.py:166]: 是的，用 datasets.load_iris() 的前 101 个数据就会出现 var < 0 的情况。原因是第三类只有一个数据，方差本应该是 0 ，但由于这里方差计算方法的原因，其掉精度可能会使其 var < 0

**Comment by nantsing** [comment]
@deadlywing

**Review by deadlywing — APPROVED** [review]
LGTM
